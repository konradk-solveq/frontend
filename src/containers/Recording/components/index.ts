import Metrics from './metrics/Metrics';
import FullMetrics from './metrics/FullMetrics';
import PopUpHint from './PopUpHint';

export {Metrics, FullMetrics, PopUpHint};
