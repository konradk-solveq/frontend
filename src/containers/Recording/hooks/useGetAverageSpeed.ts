import {useContext, useEffect, useRef} from 'react';

import {CounterDataContext} from '@pages/main/recording/counter/context/counterContext';
import {getAverageSpeedFromDistanceAndTime} from '@utils/speed';

const useGetAverageSpeed = (startTime?: Date, started?: boolean) => {
    const previusAverageSpeedRef = useRef('0.0');
    const odometer = useContext(CounterDataContext).trackerData?.odometer || 0;

    const pauseTime = useContext(CounterDataContext).pauseTime;
    const st = startTime ? Date.parse(startTime.toUTCString()) : 0;

    const averageSpeed =
        getAverageSpeedFromDistanceAndTime(odometer, st, pauseTime) || '0.0';

    useEffect(() => {
        if (started) {
            previusAverageSpeedRef.current = averageSpeed;
        }
    }, [started, averageSpeed]);

    return started ? averageSpeed : previusAverageSpeedRef.current;
};

export default useGetAverageSpeed;
