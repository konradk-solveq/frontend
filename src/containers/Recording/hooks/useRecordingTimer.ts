import {useCallback, useEffect, useRef, useState} from 'react';

import useAppState from '@hooks/useAppState';
import {convertToCounterFormat} from '@utils/dateTime';

const useRecordingTimer = (
    time?: Date,
    isRunning?: boolean,
    pauseTime?: number,
) => {
    const interval = useRef<NodeJS.Timeout | null>(null);
    const timerStartedRef = useRef(false);
    const pauseTimeRef = useRef(0);

    const [currentTime, setCurrentTime] = useState(0);

    const {appStateVisible} = useAppState();
    const [previousState, setPrevoiusState] = useState('active');

    useEffect(() => {
        setPrevoiusState(appStateVisible);
    }, [appStateVisible]);

    useEffect(() => {
        if (pauseTime) {
            pauseTimeRef.current = pauseTime;
        }
    }, [pauseTime]);

    const startTimer = () => {
        setCurrentTime(Date.now());

        timerStartedRef.current = true;
    };

    const setTime = useCallback(() => {
        if (time) {
            startTimer();
        }
    }, [time]);

    useEffect(() => {
        if (
            appStateVisible === 'active' &&
            previousState === 'background' &&
            timerStartedRef.current
        ) {
            setTime();

            setPrevoiusState('active');
        }
    }, [appStateVisible, previousState, setTime]);

    useEffect(() => {
        if (isRunning) {
            startTimer();

            interval.current = setInterval(() => {
                setCurrentTime(prevTime => prevTime + 1000);
            }, 1000);
        } else {
            timerStartedRef.current = false;
        }

        return () => {
            clearInterval(interval.current as NodeJS.Timeout);
        };
    }, [isRunning]);

    const convertedTime = convertToCounterFormat(
        currentTime,
        pauseTimeRef.current,
        time,
    );

    /**
     * Returns hours and minutes as one value
     * and seconds as secoond
     */
    return convertedTime;
};

export default useRecordingTimer;
