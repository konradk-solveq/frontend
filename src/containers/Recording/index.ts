import ThankYouPageContainer from './ThankYouPageContainer';

export {ThankYouPageContainer};
