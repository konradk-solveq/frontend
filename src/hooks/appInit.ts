import {useEffect, useState} from 'react';
import {State} from 'react-native-background-geolocation-android';

import {I18n} from '../../I18n/I18n';
import {initCrashlytics} from '../utils/crashlytics';
import {initBGeolocalization, cleanUp} from '../utils/geolocation';
import {useAppDispatch, useAppSelector} from './redux';
import {appSyncData, clearAppError} from '../storage/actions/app';
import {authTokenSelector, userIdSelector} from '../storage/selectors/auth';
import {
    appErrorSelector,
    isOnlineAppStatusSelector,
    syncAppSelector,
} from '../storage/selectors';
import {setAutorizationHeader} from '../api/api';

const useAppInit = () => {
    const trans: any = I18n.t('Geolocation.notification');
    const dispatch = useAppDispatch();
    const isOnline = useAppSelector<boolean>(isOnlineAppStatusSelector);
    const userId = useAppSelector<string>(userIdSelector);
    const authToken = useAppSelector<string>(authTokenSelector);
    const userName = useAppSelector<string>(state => state.user.userName);
    const syncStatus = useAppSelector(syncAppSelector);
    const error = useAppSelector(
        appErrorSelector,
    ); /* TODO: check all errors from sync requests */

    const [geolocationState, setGeolocationState] = useState<State>();
    const [crashlyticsInitialized, setCrashlyticsInitialized] = useState(false);
    const [dataInitialized, setDataInitialized] = useState(false);

    const clearAppSyncError = () => {
        dispatch(clearAppError());
    };

    const synchData = async () => {
        dispatch(appSyncData());
        setDataInitialized(true);
    };

    /* Set token in axios instance */
    useEffect(() => {
        if (authToken) {
            setAutorizationHeader(authToken);
        }
    }, [authToken]);

    useEffect(() => {
        /* Logs will be send after app restarted */
        initCrashlytics(userName, userId);
        setCrashlyticsInitialized(true);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    useEffect(() => {
        const init = async () => {
            const geolocation = await initBGeolocalization(trans.title);
            setGeolocationState(geolocation);
        };
        init();

        return () => {
            cleanUp();
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    useEffect(() => {
        if (isOnline && authToken) {
            synchData();
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [authToken]);

    return {
        geolocationState,
        crashlyticsInitialized,
        dataInitialized,
        isOnline,
        syncStatus,
        error,
        clearAppSyncError,
    };
};

export default useAppInit;
