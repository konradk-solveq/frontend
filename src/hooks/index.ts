import useLocationProvider from './useLocationProvider';

export {useLocationProvider};
