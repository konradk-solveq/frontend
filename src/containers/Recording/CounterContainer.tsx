import React, {useCallback, useEffect, useRef, useState} from 'react';
import {GestureResponderEvent, StyleSheet, View} from 'react-native';
import Animated, {
    useAnimatedStyle,
    useSharedValue,
    withTiming,
} from 'react-native-reanimated';

import {MIN_ROUTE_LENGTH} from '@helpers/global';
import {useAppSelector} from '@hooks/redux';
import {trackerActiveSelector} from '@storage/selectors';
import {
    trackerRecordingStateSelector,
    trackerStartTimeSelector,
} from '@storage/selectors/routes';
import {RecordingStateT} from '@storage/reducers/routes';
import {useMergedTranslation} from '@utils/translations/useMergedTranslation';
import {StateType} from '@type/components/bottomModal';
import {
    getFHorizontalPx,
    getFVerticalPx,
} from '@theme/utils/appLayoutDimensions';
import {
    appContainerHorizontalMargin,
    BOTTOM_TAB_HEIGHT,
    screenHeight,
} from '@theme/commonStyle';
import colors from '@theme/colors';

import {BottomModal} from '@components/modals';
import {HorizontalSpacer} from '@components/divider';
import {
    FillUpButton,
    PrimaryButton,
    SecondaryButton,
} from '@components/buttons';

import {FullMetrics, Metrics} from '@containers/Recording/components';
import {PopUpHint} from '@containers/Recording/components';

export const BOTTOM_MODAL_HEIGHT = 303;
export const BOTTOM_MODAL_HEIGHT_AFTER_START = 223;
const BOTTOM_SPACE_AFTER_START = getFVerticalPx(50);
const BOTTOM_SPACE = getFVerticalPx(50) + BOTTOM_TAB_HEIGHT;

interface IProps {
    onPressStart: (e: GestureResponderEvent) => void;
    onPressPauseResume: (
        e: GestureResponderEvent,
        shouldPause: boolean,
    ) => void;
    onPressStop: () => void;
    totalDistance?: number;
    topSpace?: number;
    testID?: string;
}

const CounterContainer: React.FC<IProps> = ({
    onPressStart,
    onPressPauseResume,
    onPressStop,
    totalDistance = 0,
    topSpace = 0,
    testID = 'counter-container',
}: IProps) => {
    /**
     * Keeps information that counter has been fired
     * and `default` state should not be longer presented
     */
    const recordProccessWasStartedRef = useRef(false);
    /**
     * Recording process is active (from start to stop)
     */
    const isTrackerActive = useAppSelector(trackerActiveSelector);
    const recordingState = useAppSelector(trackerRecordingStateSelector);
    /**
     * Helps to check if timer should be running
     */
    const isRecordingActive = isTrackerActive && recordingState === 'recording';
    const trackerStartTime = useAppSelector(trackerStartTimeSelector);

    const [showPopUpHint, setShowPopUpHint] = useState(false);
    const [openBottomModal, setOpenBottomModal] = useState(false);
    const [modalHeight, setModalHeight] = useState(getFVerticalPx(300));

    useEffect(() => {
        if (
            recordingState === 'not-started' &&
            !recordProccessWasStartedRef.current
        ) {
            setModalHeight(getFVerticalPx(BOTTOM_MODAL_HEIGHT));
        } else if (recordingState === 'recording') {
            recordProccessWasStartedRef.current = true;
            setModalHeight(getFVerticalPx(BOTTOM_MODAL_HEIGHT_AFTER_START));
        }
    }, [recordingState]);

    const closeContentOpacity = useSharedValue(1);
    const openContentOpacity = useSharedValue(0);
    const bottomSpaceHeight = useSharedValue(BOTTOM_SPACE);
    const closeContentAnimation = useAnimatedStyle(() => ({
        opacity: withTiming(closeContentOpacity.value, {
            duration: closeContentOpacity.value < 0 ? 150 : 500,
        }),
    }));
    const openContentAnimation = useAnimatedStyle(() => ({
        opacity: withTiming(openContentOpacity.value, {
            duration: openContentOpacity.value >= 0.2 ? 500 : 150,
        }),
    }));
    const bottomSpaceAnimation = useAnimatedStyle(() => ({
        height: withTiming(bottomSpaceHeight.value, {
            duration: 200,
        }),
    }));

    /**
     * Adjust height of bottom modal based on the state of the recording
     * and the bottom tabBar visibility
     */
    useEffect(() => {
        bottomSpaceHeight.value = !recordProccessWasStartedRef.current
            ? BOTTOM_SPACE
            : BOTTOM_SPACE_AFTER_START;
    }, [bottomSpaceHeight, recordingState]);

    /**
     * Calculates opacity based on modal height
     */
    const updateModalContentVisibility = useCallback(
        (type: StateType, position?: number) => {
            switch (type) {
                case 'closed':
                    if (openBottomModal) {
                        setOpenBottomModal(false);
                    }

                    closeContentOpacity.value = 1;
                    openContentOpacity.value = 0;
                    break;
                case 'open':
                    if (!openBottomModal) {
                        setOpenBottomModal(true);
                    }

                    closeContentOpacity.value = 0;
                    openContentOpacity.value = 1;
                    break;
                case 'active':
                    if (position) {
                        const opacityToChange = Math.abs(position) / 1000;
                        if (position > screenHeight / 2) {
                            closeContentOpacity.value =
                                closeContentOpacity.value > 0
                                    ? closeContentOpacity.value -
                                      opacityToChange
                                    : 0;
                            openContentOpacity.value =
                                openContentOpacity.value < 1
                                    ? openContentOpacity.value + opacityToChange
                                    : 1;
                        } else {
                            closeContentOpacity.value =
                                closeContentOpacity.value < 1
                                    ? closeContentOpacity.value +
                                      opacityToChange
                                    : 1;
                            openContentOpacity.value =
                                openContentOpacity.value > 0
                                    ? openContentOpacity.value - opacityToChange
                                    : 0;
                        }
                    }
                    break;
                default:
                    break;
            }
        },
        [closeContentOpacity, openContentOpacity, openBottomModal],
    );

    const onPressStopHandler = (finishedAnimation?: boolean) => {
        /**
         * When route is not long enough, user can not stop recording instantly
         */
        if (!finishedAnimation && totalDistance >= MIN_ROUTE_LENGTH) {
            setShowPopUpHint(true);
            return;
        }
        setShowPopUpHint(false);
        onPressStop();
    };

    return (
        <View style={styles.container} pointerEvents="box-none" testID={testID}>
            <BottomModal
                show
                openModal={openBottomModal}
                openModalHeight={modalHeight}
                isReactive
                isSwipeable
                autoClose
                openDuration={500}
                closeDuration={500}
                drawUnderStatusBar
                onChangeState={updateModalContentVisibility}
                style={styles.modal}>
                <View>
                    <Animated.View
                        style={[
                            styles.metricsContainer,
                            closeContentAnimation,
                        ]}>
                        <Metrics
                            onOpenPress={() => {
                                setOpenBottomModal(true);
                                closeContentOpacity.value = 0;
                                openContentOpacity.value = 1;
                            }}
                            startTime={trackerStartTime}
                            started={isRecordingActive}
                        />
                    </Animated.View>
                    <Animated.View style={[{}, openContentAnimation]}>
                        <FullMetrics
                            onOpenPress={() => {
                                setOpenBottomModal(false);
                                closeContentOpacity.value = 1;
                                openContentOpacity.value = 0;
                            }}
                            startTime={trackerStartTime}
                            started={isRecordingActive}
                            topSpace={topSpace}
                        />
                    </Animated.View>

                    <HorizontalSpacer height={25} />
                </View>
            </BottomModal>

            <View style={styles.startButtonContainer}>
                <ActionButtons
                    onPressStart={onPressStart}
                    onPressPauseResume={onPressPauseResume}
                    onPressStop={onPressStopHandler}
                    started={recordProccessWasStartedRef.current}
                    recordingState={recordingState}
                    testID={`${testID}-action-buttons`}
                />

                <Animated.View style={bottomSpaceAnimation} />
                {recordingState && recordingState !== 'not-started' && (
                    <PopUpHint
                        show={showPopUpHint}
                        onPress={() => setShowPopUpHint(false)}
                    />
                )}
            </View>
        </View>
    );
};

interface IPropsActionButtons {
    started?: boolean;
    onPressStart: (e: GestureResponderEvent) => void;
    onPressPauseResume: (e: GestureResponderEvent, resume: boolean) => void;
    onPressStop: (finishedAnimation?: boolean) => void;
    recordingState: RecordingStateT;
    testID?: string;
}

const ActionButtons: React.FC<IPropsActionButtons> = React.memo(
    ({
        started = false,
        onPressStart,
        onPressPauseResume,
        onPressStop,
        recordingState,
        testID = 'action-buttons',
    }: IPropsActionButtons) => {
        const {t} = useMergedTranslation('MainCounter.actions');
        const [isProcessing, setIsProcessing] = useState(false);

        const onPressStartHandler = useCallback(
            (e: GestureResponderEvent) => {
                setIsProcessing(true);
                onPressStart(e);
            },
            [onPressStart],
        );

        const onPressPauseResumeHandler = useCallback(
            (e: GestureResponderEvent) => {
                setIsProcessing(true);
                onPressPauseResume(e, recordingState !== 'paused');
            },
            [onPressPauseResume, recordingState],
        );

        const onPressStopHandler = useCallback(
            (finishedAnimation?: boolean) => {
                setIsProcessing(true);
                onPressStop(finishedAnimation);
            },
            [onPressStop],
        );

        useEffect(() => {
            setIsProcessing(false);
        }, [started]);

        return !started ? (
            <PrimaryButton
                text={t('startRecording')}
                withoutShadow
                withLoader={isProcessing}
                onPress={onPressStartHandler}
                testID={`${testID}-start-button`}
            />
        ) : (
            <View style={styles.row}>
                <SecondaryButton
                    text={
                        recordingState === 'paused'
                            ? t('resumeRecording')
                            : t('pauseRecording')
                    }
                    onPress={onPressPauseResumeHandler}
                    style={styles.actionButton}
                    testID={`${testID}-pause-resume-button`}
                />
                <FillUpButton
                    text={t('stopRecording')}
                    onReleaseAction={() => onPressStopHandler(false)}
                    onFilledAction={() => onPressStopHandler(true)}
                    containerStyle={styles.actionButton}
                    testID={`${testID}-stop-button`}
                />
            </View>
        );
    },
);

const styles = StyleSheet.create({
    container: {
        position: 'absolute',
        height: '100%',
        width: '100%',
        zIndex: 10,
    },
    modal: {
        backgroundColor: colors.whiteGrey,
        paddingHorizontal: appContainerHorizontalMargin,
    },
    startButtonContainer: {
        position: 'absolute',
        bottom: 0,
        width: '100%',
        alignItems: 'center',
        zIndex: 11,
    },
    metricsContainer: {
        position: 'absolute',
        left: 0,
        right: 0,
        zIndex: 1,
    },
    actionButton: {
        width: getFHorizontalPx(171),
    },
    row: {
        flexDirection: 'row',
        paddingHorizontal: appContainerHorizontalMargin,
        width: '100%',
        justifyContent: 'space-between',
    },
});

export default React.memo(CounterContainer);
