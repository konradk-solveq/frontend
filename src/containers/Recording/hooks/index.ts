import useRecordingTimer from './useRecordingTimer';
import useGetAverageSpeed from './useGetAverageSpeed';

export {useGetAverageSpeed, useRecordingTimer};
