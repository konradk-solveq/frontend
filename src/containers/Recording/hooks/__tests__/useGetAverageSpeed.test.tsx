import {renderHook, cleanup} from '@testing-library/react-hooks';

import {
    hookWrapperCounterDataProvider,
    HookWrapperProps,
} from '@jestUtils/render';
import asyncEvent from '@jestUtils/asyncEvent';
import {useGetAverageSpeed} from '@containers/Recording/hooks';

const START_DATE = '2022-06-07T06:00:00.000Z';
const DEFAULT_SPEED = '0.0';
const TRACKER_DATA = {
    distance: '3,41',
    speed: '10.0',
    averageSpeed: '6.0',
    odometer: 3406.699951171875,
    coords: {
        lat: 50.691728031513534,
        lon: 17.79613619421019,
    },
    timestamp: 1654582433,
};

describe('[useGetAverageSpeed] - containers/Recording/hooks', () => {
    describe('Permission not granted', () => {
        beforeEach(() => {
            jest.spyOn(global.Date, 'now').mockReturnValue(1654582433000);
        });

        it('Should return average speed when given "startTime" and recodring has been "started"', async () => {
            const {result, waitForNextUpdate} = renderHook(
                () => useGetAverageSpeed(new Date(START_DATE), true),
                {
                    wrapper: hookWrapperCounterDataProviderWithData,
                },
            );

            await asyncEvent(async () => {
                await waitForNextUpdate();
            });

            expect(result.current).toMatchSnapshot();
        }, 5000);

        it('Should return default speed when given "startTime" is undefined and recodring has been "started"', async () => {
            const {result, waitForNextUpdate} = renderHook(
                () => useGetAverageSpeed(undefined, true),
                {
                    wrapper: hookWrapperCounterDataProviderWithData,
                },
            );

            await asyncEvent(async () => {
                await waitForNextUpdate();
            });

            expect(result.current).toEqual(DEFAULT_SPEED);
        }, 5000);

        it('Should return default speed when given "startTime" and recodring has not been "started"', async () => {
            const {result, waitForNextUpdate} = renderHook(
                () => useGetAverageSpeed(new Date(START_DATE)),
                {
                    wrapper: hookWrapperCounterDataProviderWithData,
                },
            );

            await asyncEvent(async () => {
                await waitForNextUpdate();
            });

            expect(result.current).toEqual(DEFAULT_SPEED);
        }, 5000);

        afterEach(() => {
            jest.restoreAllMocks();
            jest.clearAllMocks();
            cleanup();
        });
    });
});

const hookWrapperCounterDataProviderWithData = ({
    children,
    initState,
}: HookWrapperProps) =>
    hookWrapperCounterDataProvider(
        {children, initState},
        {trackerData: TRACKER_DATA, pauseTime: 0},
    );
