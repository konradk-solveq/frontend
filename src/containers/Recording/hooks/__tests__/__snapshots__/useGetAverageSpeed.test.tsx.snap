// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[useGetAverageSpeed] - containers/Recording/hooks Permission not granted Should return average speed when given "startTime" and recodring has been "started" 1`] = `"14.7"`;
